integrations/tensorflow/e2e/keras/layers/layers_test.py

```diff
@@ -16,528 +16,463 @@
+LAYERS_TO_TF_UNSUPPORTED_NON_DEFAULT_KWARGS = {
+    "AveragePooling2D": ["data_format"],
+    "Conv1D": ["data_format"],
+    "Conv1DTranspose": ["data_format", "dilation_rate"],
+    "Conv2D": ["data_format"],
+    "Conv3D": ["data_format"],
+    "Conv3DTranspose": ["data_format"],
+    "LocallyConnected1D": ["padding"],
+    "LocallyConnected2D": ["padding"],
+}
+
+# Some layers have kwargs which cannot both have non-default values.
+LAYERS_TO_MUTUALLY_EXCLUSIVE_KWARGS = {
+    "Conv1D": ["strides", "dilation_rate"],
+    "Conv2D": ["strides", "dilation_rate"],
+    "Conv2DTranspose": ["strides", "dilation_rate"],
+    "Conv3D": ["strides", "dilation_rate"],
+    "ConvLSTM2D": ["strides", "dilation_rate"],
+}
+
+
+def get_default_kwargs_values(layer: str) -> Dict[str, Any]:
+  """Gets the default kwargs for a tf.keras.layers layer."""
+  layer_class = getattr(tf.keras.layers, layer)
+  layer_parameters = inspect.signature(layer_class.__init__).parameters
+  kwargs_to_default_values = {
+      kwarg: value.default
+      for kwarg, value in layer_parameters.items()
+      if value.default is not inspect.Parameter.empty
+  }
+  return kwargs_to_default_values
+
+
+def _flatten_sequence(sequence: Union[List[Any], Tuple[Any]]) -> List[Any]:
+  elements = []
+  for element in sequence:
+    if isinstance(element, (list, tuple)):
+      elements += _flatten_sequence(element)
+    else:
+      elements.append(element)
+  return elements
+
+
+def keras_default_equal(value: Any, default: Any) -> bool:
+  """Return True if 'value' is equal to 'default' as Keras kwargs."""
+  # This allows us to reuse variables like CONV_KWARGS_TO_VALUES.
+  # Treat splats as equal (e.g. 1 == (1, 1, 1, (1, 1))).
+  if (isinstance(default, Sequence) and
+      all([value == d for d in _flatten_sequence(default)])):
+    return True
+  # Keras' 1D Convs layers use 'channels_last' as a default for data_format
+  # while the 2D and 3D variants use 'None' as the default.
+  if value in [None, "channels_last"] and default in [None, "channels_last"]:
+    return True
+  return value == default
+
+
+def get_non_default_kwargs(
+    layer: str, unit_test_spec: tf_test_utils.UnitTestSpec) -> List[str]:
+  """Returns all non-default unrequired kwargs in unit_test_spec."""
+  kwargs_to_defaults = get_default_kwargs_values(layer)
+  non_default_kwargs = []
+  for kwarg, value in unit_test_spec.kwargs.items():
+    if (kwarg in kwargs_to_defaults and
+        not keras_default_equal(value, kwargs_to_defaults[kwarg])):
+      non_default_kwargs.append(kwarg)
+  return non_default_kwargs
+
+
+def unsupported_by_tf(layer: str,
+                      unit_test_spec: tf_test_utils.UnitTestSpec) -> bool:
+  """True if unit_test_spec specifies tf-unsupported non-default kwargs."""
+  if layer in LAYERS_TO_TF_UNSUPPORTED_NON_DEFAULT_KWARGS:
+    unsupported_kwargs = LAYERS_TO_TF_UNSUPPORTED_NON_DEFAULT_KWARGS[layer]
+    non_default_kwargs = get_non_default_kwargs(layer, unit_test_spec)
```

Comment:
Seems redundant to filter for non-default here. Even if it's the default it would be a problem if it wasn't supported by TF, no?

Reply:
If a tf layer has required arguments that don't work for any value, I would really hope they wouldn't include it in their API.  But `non_default_kwargs` represents kwargs in `unit_test_spec` which have non-default values, so we can't replace it with 
```python
return set(mutually_exclusive_kwargs).issubset(set(unit_test_spec.kwargs))
```
because then we wouldn't test the default values either (and wouldn't test anything).